### PR TITLE
File type inference (as it were.)

### PR DESCRIPTION
**Pull request by paulp**

### CODE REVIEW

**Comment by paulp** [comment]
I'm going to merge, comment still welcome.

File: src/main/scala/data.scala [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+// presence or absence of a parameter might be significant to the processing of a
+// media-type, depending on its definition within the media type registry.
+//
+// Note that some older HTTP applications do not recognize media type parameters.
+// When sending data to older HTTP applications, implementations SHOULD only use
+// media type parameters when they are required by that type/subtype definition.
+//
+// Media-type values are registered with the Internet Assigned Number Authority
+// (IANA [19]). The media type registration process is outlined in RFC 1590 [17].
+// Use of non-registered media types is discouraged.
```
Comment (EECOLOR) [inline src/main/scala/data.scala:25]: we don't need these last two paragraphs

File: src/main/scala/data.scala [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+// https://www.w3.org/Protocols/rfc2616/rfc2616-sec3.html#sec3.7
+//
+// media-type     = type "/" subtype *( ";" parameter )
+// type           = token
+// subtype        = token
+//
+// Parameters MAY follow the type/subtype in the form of attribute/value pairs
+// (as defined in section 3.6).
+//
+// The type, subtype, and parameter attribute names are case-insensitive.
```
Comment (EECOLOR) [inline src/main/scala/data.scala:12]: This is not reflected in the code

File: src/main/scala/data.scala [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+// Use of non-registered media types is discouraged.
+
+final case class MediaType(`type`: String, subtype: String, attrs: Map[String, String]) {
+  def charset = attrs.getOrElse("charset", "<none>")
+  override def toString = s"%s/%s".format(`type`, subtype)
+}
+object MediaType {
+  val NoMediaType = new MediaType("<none>", "<none>", Map())
+
+  val Extract = """^([^/]+)/([^\s;]+)[;]\s*(.*)$""".r
```
Comment (EECOLOR) [inline src/main/scala/data.scala:34]: > Linear white space (LWS) MUST NOT be used between the type and subtype

The regex allows it. If you don't care about these sort of things, remove the comment :-)

File: src/main/scala/data.scala [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+  private def mkPair(s: String) = s indexOf '=' match {
```
Comment (EECOLOR) [inline src/main/scala/data.scala:36]: I personally like it when people write code top to bottom. I think it's far superior when reading code to be only confronted with implementation details once the context is present. Although most people agree with me once I discuss it with them. It's still very much rooted in the 'define before you use' style that was required way back.

File: src/main/scala/data.scala [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+// Media-type values are registered with the Internet Assigned Number Authority
+// (IANA [19]). The media type registration process is outlined in RFC 1590 [17].
+// Use of non-registered media types is discouraged.
+
+final case class MediaType(`type`: String, subtype: String, attrs: Map[String, String]) {
+  def charset = attrs.getOrElse("charset", "<none>")
+  override def toString = s"%s/%s".format(`type`, subtype)
+}
+object MediaType {
+  val NoMediaType = new MediaType("<none>", "<none>", Map())
```
Comment (EECOLOR) [inline src/main/scala/data.scala:32]: I'd put the `"<none>"` in a value, that allows you to later check on it. Another option is to be more precise with the types and create a type that can be empty (or use an `Option`)

File: src/main/scala/jio/package.scala [hunk 6]
```diff
@@ -60,26 +43,55 @@ package object jio extends JioFiles {
...
-
     private def withWriteChannel[A](code: jnc.FileChannel => A): A = {
-      val channel = jnc.FileChannel.open(p, jnf.StandardOpenOption.WRITE)
+      val channel = jnc.FileChannel.open(path, jnf.StandardOpenOption.WRITE)
       try code(channel)
       finally channel.close()
     }
   }
 
+  trait Pathish[Rep] {
```
Comment (EECOLOR) [inline src/main/scala/jio/package.scala:62]: I like this word :-)

File: src/main/scala/jio/package.scala [hunk 7]
```diff
@@ -28,30 +21,20 @@ package object jio extends JioFiles {
...
-    def readSymbolicLink                 = Files readSymbolicLink p
-    def allBytes                         = Files readAllBytes p
-    def list: Vector[Path]               = (Files list p).toVector
-    def filename: String                 = p.getFileName.toString
-    def blockSize: Long                  = 512 // FIXME
-    def blockCount: Long                 = (attributes.size + blockSize - 1) / blockSize
-    def size: Long                       = attributes.size
-    def atime: Long                      = attributes.lastAccessTime.toMillis
-    def mtime: Long                      = attributes.lastModifiedTime.toMillis
+  implicit class FileOps(val f: File) extends Pathish[File] {
```
Comment (EECOLOR) [inline src/main/scala/jio/package.scala:24]: My gut feeling tells me we don't need this one anymore. Just a comment for when one of us feels like removing it.

File: src/main/scala/package.scala [hunk 8]
```diff
@@ -19,6 +19,18 @@ package object suffuse {
...
   def tryFuse(body: => Unit): Int       = Try(body) fold (_.toErrno, _ => eok)
 
+  // For example statsBy(path("/usr/bin").ls)(_.mediaType.subtype)
+  //
+  // 675   octet-stream
+  // 96    x-shellscript
+  // 89    x-perl
+  // 26    x-c
+  // ...
+  def statsBy[A, B](xs: Seq[A])(f: A => B): Unit = {
```
Comment (EECOLOR) [inline src/main/scala/package.scala:29]: This is just for debugging purposes?

File: src/main/scala/jio/package.scala (hunk 7 above)
Comment (paulp) [inline src/main/scala/jio/package.scala:24]: I can tell you from experience that File turns up all over the place. It's just a lot less hassle if you can treat File and Path as interchangeable. There isn't any relevant distinction between them, but you will always need one or the other. Outright implicit conversions have a way of burning.

File: src/main/scala/package.scala (hunk 8 above)
Comment (paulp) [inline src/main/scala/package.scala:29]: Plus or minus. It's ridiculously handy and I get tired of needing it in every project.

File: src/main/scala/jio/package.scala (hunk 7 above)
Comment (EECOLOR) [inline src/main/scala/jio/package.scala:24]: I hear what you are saying. Another reason to get rid of it is that I moved to `Files` for quite a few features that did not behave correctly (latest was `renameTo`)

File: src/main/scala/package.scala (hunk 8 above)
Comment (EECOLOR) [inline src/main/scala/package.scala:29]: Cool

**Pull request merged**